### PR TITLE
[EAGLE-2][EAGLE-24][EAGLE-50][EAGLE-52]Add eagle offline metric collection topology and do online balance partition based on the statistic metric

### PR DESCRIPTION
**Pull request by sunlibin**

### CODE REVIEW

**Comment by haoch** [comment]
@sunlibin This pull request has conflict with master branch that must be resolved, please resolve firstly.

**Comment by sunlibin** [comment]
@haoch, just fix the conflicts

**Comment by haoch** [comment]
LGTM, need clean commit history

File: eagle-security/eagle-metric-collection/src/main/java/org/apache/eagle/metric/kafka/EagleMetricCollectorMain.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            @Override
+            public List<Object> deserialize(byte[] ser) {
+                Object tmp = deserializer.deserialize(ser);
+                Map<String, Object> map = (Map<String, Object>)tmp;
+                if(tmp == null) return null;
+                return Arrays.asList(map.get("user"), map.get("timestamp"));
+            }
+        };
+
+        KafkaSourcedSpoutProvider kafkaMessageSpoutProvider = new KafkaSourcedSpoutProvider() {
```
Comment (haoch) [inline eagle-security/eagle-metric-collection/src/main/java/org/apache/eagle/metric/kafka/EagleMetricCollectorMain.java]: Never pull too long anonymous method in main entry method, especially when the design is not very clean

File: eagle-security/eagle-security-hdfs-auditlog/src/main/java/org/apache/eagle/security/auditlog/HdfsAuditLogProcessorMain.java [hunk 2]
```diff
@@ -37,34 +44,72 @@
...
+    }
+
+	public static void main(String[] args) throws Exception{
+        Config config = new ConfigOptionParser().load(args);
+        LOG.info("Config class: " + config.getClass().getCanonicalName());
+        if(LOG.isDebugEnabled()) LOG.debug("Config content:"+config.root().render(ConfigRenderOptions.concise()));
+
+        StormExecutionEnvironment env = ExecutionEnvironmentFactory.getStorm(config);
+        KafkaSourcedSpoutProvider provider = createProvider(config);
+        Boolean balancePartition = config.hasPath("eagleProps.balancePartitionEnabled") ? config.getBoolean("eagleProps.balancePartitionEnabled") : false;
```
Comment (haoch) [inline eagle-security/eagle-security-hdfs-auditlog/src/main/java/org/apache/eagle/security/auditlog/HdfsAuditLogProcessorMain.java]: Such kind of config key is not very good

**Comment by haoch** [comment]
1) The change here https://github.com/apache/incubator-eagle/blob/master/eagle-core/eagle-data-process/eagle-stream-process-api/src/main/java/org/apache/eagle/dataproc/impl/storm/kafka/KafkaSourcedSpoutScheme.java#L63 breaks the previous features
2) Keep default implementation for kafka provider instead of forcing developer to declare schema

**Comment by sunlibin** [comment]
@haoch , I check the commit log, I didn't change KafkaSourcedSpoutScheme

**Comment by haoch** [comment]
@sunlibin I think the file was created for the work about user-based partition in Spout, wasn't it?

**Comment by haoch** [comment]
Just keep it, there is some other problem about the implementation like forcing developer to do too much work while using kafka and code in logic level is too complex, I am refactoring about it and will discuss in mailing list later.

**Comment by sunlibin** [comment]
@haoch, I'm afraid not, by checking the commit, we can see the code exists in the initial commit

**Comment by haoch** [comment]
@sunlibin, thanks for the info. It doesn't matter. I will fix it :-)

**Pull request merged**